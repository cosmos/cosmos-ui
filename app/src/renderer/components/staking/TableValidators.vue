<template lang="pug">
  div
    tm-data-loading(v-if="delegates.loading && sortedFilteredEnrichedDelegates.length === 0")
    tm-data-empty(v-else-if="!delegates.loading && validators.length === 0")
    data-empty-search(v-else-if="!delegates.loading && sortedFilteredEnrichedDelegates.length === 0")
    table(v-else)
      thead
        panel-sort(:sort='sort', :properties="properties")
      tbody
        li-validator(v-for='i in sortedFilteredEnrichedDelegates', :disabled="!userCanDelegate" :key='i.id' :validator='i')
</template>

<script>
import { mapGetters } from "vuex"
import num from "scripts/num"
import { includes, orderBy } from "lodash"
import Mousetrap from "mousetrap"
import LiValidator from "staking/LiValidator"
import { TmDataEmpty, TmDataLoading } from "@tendermint/ui"
import DataEmptySearch from "common/TmDataEmptySearch"
import { calculateTokens } from "scripts/common"
import ModalSearch from "common/TmModalSearch"
import PanelSort from "staking/PanelSort"
import VmToolBar from "common/VmToolBar"
export default {
  name: `table-validators`,
  props: [`validators`],
  components: {
    LiValidator,
    TmDataEmpty,
    DataEmptySearch,
    TmDataLoading,
    ModalSearch,
    PanelSort,
    VmToolBar
  },
  data: () => ({
    num: num,
    query: ``,
    sort: {
      property: `percent_of_vote`,
      order: `desc`
    }
  }),
  computed: {
    ...mapGetters([
      `delegates`,
      `delegation`,
      `filters`,
      `committedDelegations`,
      `config`,
      `user`,
      `connected`,
      `bondingDenom`,
      `keybase`
    ]),
    address() {
      return this.user.address
    },
    somethingToSearch() {
      return !!this.validators.length
    },
    vpTotal() {
      return this.validators
        .slice(0)
        .map(v => {
          v.voting_power = v.voting_power ? Number(v.voting_power) : 0
          return v
        })
        .sort((a, b) => b.voting_power - a.voting_power)
        .slice(0, 100)
        .reduce((sum, v) => sum + v.voting_power, 0)
    },
    enrichedDelegates() {
      return this.validators.map(v =>
        Object.assign({}, v, {
          small_moniker: v.description.moniker.toLowerCase(),
          percent_of_vote: num.percent(v.voting_power / this.vpTotal),
          your_votes: this.num.full(
            calculateTokens(v, this.committedDelegations[v.id])
          ),
          keybase: this.keybase[v.description.identity]
        })
      )
    },
    sortedFilteredEnrichedDelegates() {
      let query = this.filters.delegates.search.query || ``
      let sortedEnrichedDelegates = orderBy(
        this.enrichedDelegates.slice(0),
        [this.sort.property, `small_moniker`],
        [this.sort.order, `asc`]
      )
      if (this.filters.delegates.search.visible) {
        return sortedEnrichedDelegates.filter(i =>
          includes(JSON.stringify(i).toLowerCase(), query.toLowerCase())
        )
      } else {
        return sortedEnrichedDelegates
      }
    },
    userCanDelegate() {
      return this.user.atoms > 0 && this.delegation.loadedOnce
    },
    properties() {
      return [
        {
          title: `Moniker`,
          value: `small_moniker`,
          tooltip: `The validator's moniker`,
          class: `name`
        },
        {
          title: `Delegated ${this.bondingDenom}`,
          value: `your_votes`,
          tooltip: `Number of ${
            this.bondingDenom
          } you have delegated to the validator`,
          class: `your-votes`
        },
        {
          title: `Rewards`,
          value: `your_rewards`, // TODO: use real rewards
          tooltip: `Rewards of ${
            this.bondingDenom
          } you have gained from the validator`,
          class: `your-rewards` // TODO: use real rewards
        },
        {
          title: `Voting Power`,
          value: `percent_of_vote`,
          tooltip: `Percentage of ${
            this.bondingDenom
          } the validator has on The Cosmos Hub`,
          class: `percent_of_vote`
        },
        {
          title: `Uptime`,
          value: `uptime`,
          tooltip: `Ratio of blocks signed within the last 10k blocks`,
          class: `uptime`
        },
        {
          title: `Commission`,
          value: `commission`,
          tooltip: `The validator's commission`,
          class: `commission`
        },
        {
          title: `Slashes`,
          value: `slashes`, // TODO: use real slashes
          tooltip: `The validator's slashes`,
          class: `slashes`
        }
      ]
    }
  },
  watch: {
    address: function(address) {
      address && this.updateDelegates()
    }
  },
  methods: {
    updateDelegates() {
      this.$store.dispatch(`updateDelegates`)
    },
    setSearch(
      bool = !this.filters[`delegates`].search.visible,
      { somethingToSearch, $store } = this
    ) {
      if (somethingToSearch) {
        $store.commit(`setSearchVisible`, [`delegates`, bool])
      }
    }
  },
  async mounted() {
    Mousetrap.bind([`command+f`, `ctrl+f`], () => this.setSearch(true))
    Mousetrap.bind(`esc`, () => this.setSearch(false))

    // XXX temporary because querying the shares shows old shares after bonding
    // this.updateDelegates()
  }
}
</script>
<style lang="stylus">
@require '~variables'

table
  border-spacing 0 0.25rem
  margin 0 0 0 2rem
  min-width
  padding 0
  table-layout auto
  counter-reset rowNumber + 1

table tr
  counter-increment rowNumber

table tr td:first-child::before
  content counter(rowNumber)
  position absolute
  font-size sm
  width 2rem
  text-align right
  color var(--dim)
  left -3rem

table th
  min-width 130px
  width 100%
  padding 0.5rem

table td
  min-width 130px
  width 100%
  padding 0 0.5rem
  position relative

  a
    display inline-block

table tr td:nth-child(3):after
  display block
  position absolute
  content ''
  height 2rem
  width 2px
  top 1.5rem
  right 2rem
  background var(--bc-dim)

</style>
