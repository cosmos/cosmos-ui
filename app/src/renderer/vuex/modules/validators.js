export default ({ node }) => {
  const state = {
    validators: [],
    loading: false,
    subscription: false,
    validatorHash: null
  }

  const mutations = {
    setValidators(state, validators) {
      state.validators = validators
    },
    updateValidator(state, index, updatedValidator) {
      state.validators[index] = updatedValidator
    },
    addValidator(state, validator) {
      state.validators.push(validator)
    },
    setValidatorHash(state, validatorHash) {
      state.validatorHash = validatorHash
    }
  }

  const actions = {
    reconnected({ state, dispatch }) {
      if (state.loading) {
        dispatch("getValidators")
      }
    },
    getValidators({ state, commit }) {
      state.loading = true
      node.rpc.validators((err, { validators } = {}) => {
        if (err) {
          commit("notifyError", {
            title: "Error fetching validator set",
            body: err.message
          })
          return
        }
        commit("setValidators", validators)
        state.loading = false
      })
    },
    maybeUpdateValidators({ state, commit, dispatch }, header) {
      let validatorHash = header.validators_hash
      if (validatorHash === state.validatorHash) return
      commit("setValidatorHash", validatorHash)
      dispatch("getValidators")
    },
    validatorUpdateSubscribe({ state, commit, dispatch }) {
      // Return if already subscribed to updates
      if (state.subscription) return

      // TODO move to common file since subscribeToBlock also uses it
      function error(err) {
        if (err.data === "already subscribed") {
          console.error("Tryed to subscribe to an already subscribed rpc event")
          return
        }
        dispatch("nodeHasHalted")
        console.error(
          `Error subscribing to new validator set updates: ${
            err.message
          } ${err.data || ""}`
        )
      }
      node.rpc.subscribe(
        { query: "tm.event = 'ValidatorSetUpdates'" },
        (err, event) => {
          if (err) return error(err)
          // https://github.com/tendermint/tendermint/pull/2161/files#diff-76d6626766e8d862e9acb34c166664b5
          for (var validator in event.data.value.validator_updates) {
            if (validator.power == 0) {
              let dumpedValIdx = state.validators.findIndex(
                v => v.pubkey == validator.pubkey
              )
              commit("updateValidator", dumpedValIdx, validator)
            } else {
              commit("addValidator", validator)
            }
          }
        }
      )
    }
  }

  return {
    state,
    mutations,
    actions
  }
}
