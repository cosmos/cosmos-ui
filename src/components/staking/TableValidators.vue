<template>
  <div>
    <table class="data-table">
      <thead>
        <PanelSort :sort="sort" :properties="properties" />
      </thead>
      <tbody>
        <LiValidator
          v-for="validator in sortedEnrichedValidators"
          :key="validator.operator_address"
          :validator="validator"
        />
      </tbody>
    </table>
  </div>
</template>

<script>
import { mapGetters } from "vuex"
import num from "scripts/num"
import { orderBy } from "lodash"
import LiValidator from "staking/LiValidator"
import PanelSort from "staking/PanelSort"
export default {
  name: `table-validators`,
  components: {
    LiValidator,
    PanelSort
  },
  props: {
    validators: {
      type: Array,
      required: true
    }
  },
  data: () => ({
    num: num,
    query: ``,
    sort: {
      property: `commission`,
      order: `asc`
    },
    rollingWindow: 10000 // param of slashing period
  }),
  computed: {
    ...mapGetters([
      `committedDelegations`,
      `delegates`,
      `session`,
      `distribution`,
      `bondDenom`,
      `keybase`,
      `pool`
    ]),
    enrichedValidators(
      {
        validators,
        delegates: { signingInfos },
        pool,
        committedDelegations,
        keybase,
        session,
        distribution,
        rollingWindow
      } = this
    ) {
      return validators.map(v => {
        const signingInfo = signingInfos[v.operator_address]
        return Object.assign({}, v, {
          small_moniker: v.description.moniker.toLowerCase(),
          percent_of_vote: v.voting_power / pool.pool.bonded_tokens,
          my_delegations:
            session.signedIn && committedDelegations[v.id] > 0
              ? committedDelegations[v.id]
              : 0,
          commission: v.commission.rate,
          keybase: keybase[v.description.identity],
          rewards:
            session.signedIn && distribution.rewards[v.operator_address]
              ? distribution.rewards[v.operator_address][this.bondDenom]
              : 0,
          uptime: signingInfo
            ? (rollingWindow - signingInfo.missed_blocks_counter) /
              rollingWindow
            : 0
        })
      })
    },
    sortedEnrichedValidators() {
      return orderBy(
        this.enrichedValidators.slice(0),
        [this.sort.property],
        [this.sort.order]
      )
    },
    properties() {
      return [
        {
          title: `Moniker`,
          value: `small_moniker`,
          tooltip: `The validator's moniker`
        },
        {
          title: `My Delegations`,
          value: `my_delegations`,
          tooltip: `Number of ${num.viewDenom(
            this.bondDenom
          )} you have delegated to this validator`
        },
        {
          title: `Rewards`,
          value: `rewards`,
          tooltip: `Rewards you have earned from this validator`
        },
        {
          title: `Voting Power`,
          value: `percent_of_vote`,
          tooltip: `Percentage of voting shares`
        },
        {
          title: `Commission`,
          value: `commission`,
          tooltip: `The fee the validator will charge from your rewards`
        },
        {
          title: `Uptime`,
          value: `uptime`,
          tooltip: `Ratio of blocks signed within the last 10k blocks`
        }
      ]
    },
    yourValidators({ committedDelegations, validators, session } = this) {
      if (!session.signedIn) {
        return
      }

      return validators.filter(
        ({ operator_address }) => operator_address in committedDelegations
      )
    }
  },
  watch: {
    address: function() {
      this.session.address && this.$store.dispatch(`updateDelegates`)
    }
  },
  mounted() {
    this.$store.dispatch(`getPool`)
    this.$store.dispatch(`updateDelegates`)
  }
}
</script>
