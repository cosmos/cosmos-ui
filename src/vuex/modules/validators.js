import * as Sentry from "@sentry/browser"
import Vue from "vue"
import { coinsToObject } from "scripts/common.js"

export default ({ node }) => {
  const emptyState = {
    validators: [],
    loading: false,
    loaded: false,
    error: null,
    validatorHash: null,
    distributionInfo: {},
    rewards: {}
  }
  const state = JSON.parse(JSON.stringify(emptyState))

  const mutations = {
    setValidators(state, validators) {
      state.validators = validators
    },
    setValidatorHash(state, validatorHash) {
      state.validatorHash = validatorHash
    },
    setValidatorDistributionInfo(state, { validatorAddr, info }) {
      Vue.set(state.distributionInfo, validatorAddr, info)
    },
    setValidatorRewards(state, { validatorAddr, rewards }) {
      Vue.set(state.rewards, validatorAddr, rewards)
    }
  }

  const actions = {
    reconnected({ state, dispatch }) {
      if (state.loading) {
        dispatch(`getValidators`)
      }
    },
    resetSessionData({ rootState }) {
      // clear previous account state
      rootState.validators = JSON.parse(JSON.stringify(emptyState))
    },
    async getValidators({ state, commit, rootState }) {
      state.loading = true

      if (!rootState.connection.connected) return

      try {
        const validators = (await node.getValidatorSet()).validators
        state.error = null
        state.loading = false
        state.loaded = true
        commit(`setValidators`, validators)
      } catch (error) {
        commit(`notifyError`, {
          title: `Error fetching validator set`,
          body: error.message
        })
        Sentry.captureException(error)
        state.error = error
      }
    },
    async maybeUpdateValidators({ state, commit, dispatch }, header) {
      const validatorHash = header.validators_hash
      if (validatorHash === state.validatorHash) return
      commit(`setValidatorHash`, validatorHash)
      await dispatch(`getValidators`)
    },
    async getValidatorDistributionInfo({ commit }, validatorAddr) {
      state.loading = true
      try {
        let {
          self_bond_rewards,
          val_commission
        } = await node.getValidatorDistributionInformation(validatorAddr)
        self_bond_rewards = coinsToObject(self_bond_rewards)
        val_commission = coinsToObject(val_commission)
        const info = { self_bond_rewards, val_commission }
        commit(`setValidatorDistributionInfo`, { validatorAddr, info })
        state.error = null
        state.loaded = true
      } catch (error) {
        Sentry.captureException(error)
        state.error = error
      }
      state.loading = false
    },
    async getValidatorDistributionRewards({ commit }, validatorAddr) {
      state.loading = true
      try {
        const rewardsArray = await node.getValidatorRewards(validatorAddr)
        const rewards = coinsToObject(rewardsArray)
        commit(`setValidatorRewards`, { validatorAddr, rewards })
        state.error = null
        state.loaded = true
      } catch (error) {
        Sentry.captureException(error)
        state.error = error
      }
      state.loading = false
    }
  }

  return {
    state,
    mutations,
    actions
  }
}
