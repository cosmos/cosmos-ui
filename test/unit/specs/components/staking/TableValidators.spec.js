import { shallowMount } from "@vue/test-utils"
import TableValidators from "renderer/components/staking/TableValidators"
import validators from "../../store/json/validators.js"

describe(`TableValidators`, () => {
  let wrapper, $store

  const getters = {
    committedDelegations: {
      [validators[0].operator_address]: 10
    },
    session: {
      address: `address1234`,
      signedIn: true
    },
    distribution: {
      loaded: true,
      rewards: {
        cosmosvaladdr15ky9du8a2wlstz6fpx3p4mqpjyrm5ctqzh8yqw: {
          stake: 1000
        }
      }
    },
    bondDenom: `stake`,
    keybase: { [validators[0].description.identity]: `keybase` },
    pool: {
      pool: {
        bonded_tokens: 500001,
      }
    },
  }

  beforeEach(() => {
    $store = {
      commit: jest.fn(),
      dispatch: jest.fn(),
      getters: JSON.parse(JSON.stringify(getters)) // clone so we don't overwrite by accident
    }

    wrapper = shallowMount(TableValidators, {
      mocks: {
        $store
      },
      propsData: { validators }
    })
    wrapper.setData({ rollingWindow: 10000 })
  })

  it(`has the expected html structure`, async () => {
    expect(wrapper.vm.$el).toMatchSnapshot()
  })

  it(`should create an enriched validator object for a signed in user`, () => {
    expect(wrapper.vm.enrichedValidators[0].small_moniker).toBe(`mr_mounty`)
    expect(wrapper.vm.enrichedValidators[0].percent_of_vote)
      .toBe(0.27999944000112)
    expect(wrapper.vm.enrichedValidators[0].my_delegations).toBe(10)
    expect(wrapper.vm.enrichedValidators[0].commission).toBe(0)
    expect(wrapper.vm.enrichedValidators[0].keybase).toBe(`keybase`)
    expect(wrapper.vm.enrichedValidators[0].rewards).toBe(1000)
    expect(wrapper.vm.enrichedValidators[0].uptime).toBe(0.9998)
  })

  it(`should create an enriched validator object for a user who is not signed in`, () => {
    wrapper.vm.session.signedIn = false
    expect(wrapper.vm.enrichedValidators[1].my_delegations).toBe(0)
    expect(wrapper.vm.enrichedValidators[1].rewards).toBe(0)
  })

  it(`should have an uptime of 0 if no signing_info`, () => {
    expect(wrapper.vm.enrichedValidators[1].uptime).toBe(0)
  })

  it(`should sort the delegates by selected property`, () => {
    wrapper.vm.sort.property = `operator_address`
    wrapper.vm.sort.order = `desc`

    expect(
      wrapper.vm.sortedEnrichedValidators.map(x => x.operator_address)
    ).toEqual(validators.map(x => x.operator_address))

    wrapper.vm.sort.property = `operator_address`
    wrapper.vm.sort.order = `asc`

    expect(
      wrapper.vm.sortedEnrichedValidators.map(x => x.operator_address)
    ).toEqual(validators.map(x => x.operator_address).reverse())
  })

  it(`queries delegations on signin`, () => {
    const session = { address: `cosmos1address` }
    const $store = { dispatch: jest.fn() }
    TableValidators.watch.address.call({ $store, session })
    expect($store.dispatch).toHaveBeenCalledWith(`updateDelegates`)
  })

  it(`doesn't query delegations if not signed in`, () => {
    const session = { address: undefined }
    const $store = { dispatch: jest.fn() }
    TableValidators.watch.address.call({ $store, session })
    expect($store.dispatch).not.toHaveBeenCalledWith(`updateDelegates`)
  })

  it(`should filter the validators for your delegations`, () => {
    const session = { signedIn: true }
    expect(
      TableValidators.computed.yourValidators({
        committedDelegations: {
          [validators[0].operator_address]: 1,
          [validators[2].operator_address]: 2
        },
        validators,
        session
      })
    ).toEqual([validators[0], validators[2]])
  })

  it(`should not filter the validators if you're not signed in`, () => {
    const session = { signedIn: false }
    expect(
      TableValidators.computed.yourValidators({
        committedDelegations: {
          [validators[0].operator_address]: 1,
          [validators[2].operator_address]: 2
        },
        validators,
        session
      })
    ).not.toBeDefined()
  })
})
