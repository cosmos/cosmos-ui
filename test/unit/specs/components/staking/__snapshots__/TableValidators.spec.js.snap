// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableValidators has the expected html structure 1`] = `
<div>
  <table
    class="data-table"
  >
    <thead>
      <panel-sort-stub
        properties="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]"
        sort="[object Object]"
      />
    </thead>
     
    <tbody>
      <li-validator-stub
        validator="[object Object]"
      />
      <li-validator-stub
        validator="[object Object]"
      />
      <li-validator-stub
        validator="[object Object]"
      />
    </tbody>
  </table>
</div>
`;
